vm/src/builtins/pyunion.rs

```diff
@@ -0,0 +1,289 @@
+use super::genericalias;
+use crate::{
+    builtins::{PyFrozenSet, PyStr, PyStrRef, PyTuple, PyTupleRef, PyTypeRef},
+    common::hash,
+    function::IntoPyObject,
+    protocol::PyMappingMethods,
+    types::{AsMapping, Comparable, GetAttr, Hashable, Iterable, PyComparisonOp},
+    IdProtocol, PyClassImpl, PyComparisonValue, PyContext, PyObject, PyObjectRef, PyObjectView,
+    PyRef, PyResult, PyValue, TryFromObject, TypeProtocol, VirtualMachine,
+};
+use std::fmt;
+
+const CLS_ATTRS: &[&str] = &["__module__"];
+
+#[pyclass(module = "types", name = "UnionType")]
+pub struct PyUnion {
+    args: PyTupleRef,
+    parameters: PyTupleRef,
+}
+
+impl fmt::Debug for PyUnion {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("UnionObject")
+    }
+}
+
+impl PyValue for PyUnion {
+    fn class(vm: &VirtualMachine) -> &PyTypeRef {
+        &vm.ctx.types.union_type
+    }
+}
+
+#[pyimpl(with(Hashable, Comparable, AsMapping), flags(BASETYPE))]
+impl PyUnion {
+    pub fn new(args: PyTupleRef, vm: &VirtualMachine) -> Self {
+        let parameters = make_parameters(&args, vm);
+        Self { args, parameters }
+    }
+
+    #[pymethod(magic)]
+    fn repr(&self, vm: &VirtualMachine) -> PyResult<String> {
+        fn repr_item(obj: PyObjectRef, vm: &VirtualMachine) -> PyResult<String> {
+            if vm.is_none(&obj) {
+                return Ok("None".to_string());
+            }
+
+            if vm.get_attribute_opt(obj.clone(), "__origin__")?.is_some()
+                && vm.get_attribute_opt(obj.clone(), "__args__")?.is_some()
+            {
+                return Ok(obj.repr(vm)?.as_str().to_string());
+            }
+
+            match (
+                vm.get_attribute_opt(obj.clone(), "__qualname__")?
+                    .and_then(|o| o.downcast_ref::<PyStr>().map(|n| n.as_str().to_string())),
+                vm.get_attribute_opt(obj.clone(), "__module__")?
+                    .and_then(|o| o.downcast_ref::<PyStr>().map(|m| m.as_str().to_string())),
+            ) {
+                (None, _) | (_, None) => Ok(obj.repr(vm)?.as_str().to_string()),
+                (Some(qualname), Some(module)) => Ok(if module == "builtins" {
+                    qualname
+                } else {
+                    format!("{}.{}", module, qualname)
+                }),
+            }
+        }
+
+        Ok(self
+            .args
+            .as_slice()
+            .iter()
+            .map(|o| repr_item(o.clone(), vm))
+            .collect::<PyResult<Vec<_>>>()?
+            .join(" | "))
+    }
+
+    #[pyproperty(magic)]
+    fn parameters(&self) -> PyObjectRef {
+        self.parameters.as_object().to_owned()
+    }
+
+    #[pyproperty(magic)]
+    fn args(&self) -> PyObjectRef {
+        self.args.as_object().to_owned()
```

Comment:
I tried to apply this suggestion but it failed with the below logs 😿 :

```
error[E0507]: cannot move out of `self.parameters` which is behind a shared reference
  --> vm/src/builtins/pyunion.rs:79:9
   |
79 |         self.parameters.into()
   |         ^^^^^^^^^^^^^^^ move occurs because `self.parameters` has type `pyobjectrc::PyRef<PyTuple>`, which does not implement the `Copy` trait

error[E0507]: cannot move out of `self.args` which is behind a shared reference
  --> vm/src/builtins/pyunion.rs:84:9
   |
84 |         self.args.into()
   |         ^^^^^^^^^ move occurs because `self.args` has type `pyobjectrc::PyRef<PyTuple>`, which does not implement the `Copy` trait
```

Reply:
```suggestion
        self.args.clone().into()
```
Adding a `clone` will work